library/proc_macro/src/bridge/mod.rs

```diff
@@ -7,6 +7,10 @@
 //! Rust ABIs (e.g., stage0/bin/rustc vs stage1/bin/rustc during bootstrap).
 
 #![deny(unsafe_code)]
+// proc_macros anyway don't work on wasm hosts so while both sides of this bridge can
+// be built with different versions of rustc, the wasm ABI changes don't really matter.
+#![cfg_attr(bootstrap, allow(unknown_lints))]
+#![allow(wasm_c_abi)]
```

Comment:
You need to allow it at the root of the crate as you passed `CRATE_HIR_ID` to `emit_node_span_lint`.

Reply:
Ah, that's how that works. :)
Is there a better HIR ID we can pass? For the call-site check we have a MIR location, and for the callee-site check we have an `Instance`. Can one get a HIR ID from those?

Reply:
If the caller or callee is codegened in a different crate than it was defined, there is no way to get the lint level of its definition AFAIK. HIR only exists for the local crate.

Reply:
Yeah fair, but in many cases this will be local to the crate and then we should be able to get a proper HirId. I've added a commit that attempts to do that.

Reply:
Now that I think of it, having a fallback to the `CRATE_HIR_ID` would actually cause warnings to depend on the exact crate in which rustc decides to codegen functions.

Reply:
That's already the case anyway, right? With my patch we just use a better HirId when we can.

This is a mono-time check, so it fundamentally can indeed occur in other crates. There's not a lot we can do about that; ABI concerns can only be reliably checked during monomorphization.

Reply:
> That's already the case anyway, right? With my patch we just use a better HirId when we can.

Yeah, your last commit is strictly an improvement with respect to this.

> This is a mono-time check, so it fundamentally can indeed occur in other crates.

Currently all of them are hard errors, so it doesn't matter in which crate they get codegened. Only if they get codegened at all. With this PR however a crate may need to add `#![allow(...)]` because an indirect dependency happened to use this attribute. And which crates need to add it even depends on the rustc version and exact build configuration that the end user decided.